just-the-docs/just-the-docs
# Fix the navigation panel

This PR is an extension of #1243. Now that #1243 has been merged, it has significantly reduced the number of changed files.

The main change is to **fix the navigation panel**: make it _page-independent_, so that it can be generated just once and _cached_. All pages need to include all navigation links (since the navigation expanders can be used to reveal navigation links independently of the selected page) so such caching should significantly reduce the build time for larger sites.

The idea[^1] is to use JavaScript to toggle the `active` class in the navigation links above the link to the selected page, as well as in the navigation expander that reveals its children.[^2] (The author has very little experience of JS programming and the DOM API; suggestions for improvement are welcome!)

[^1]: @jhvanderschee suggested the basic idea in [a Jekyll Talk discussion of this theme's build time].

[a Jekyll Talk discussion of this theme's build time]: https://talk.jekyllrb.com/t/13-seconds-build-time-per-page/8007/6 "13 seconds build time per page, discussed on Jekyll Talk, February 2023"

[^2]: JavaScript is already used to _scroll_ the navigation panel, to ensure the link to the selected page is visible.

The PR does not involve any changes to the current CSS used to style the navigation panel. It significantly simplifies the HTML for building the navigation panel, reducing `_includes/nav.html` from 251 LOC to 75 LOC.[^3]

[^3]: Much of the reduction came from removing explanatory comments that are no longer needed; moreover, the Liquid code used to sort pages is now in a separate file.

The initial version of this PR has now been enhanced to generate page-specific CSS (for inclusion in  the `head` element) to style the navigation links in the usual way when JavaScript isn't in use. The generated rules override the current CSS. That makes the use of JavaScript [unobtrusive], and should also improve accessibility.

Profiling has shown that it can still take significant time to generate the page-specific CSS (mostly for determining the navigation path to the page). For sites with a shallow navigation hierarchy, use of JavaScript isn't needed to access all the pages, so perhaps the generation of the page-specific CSS should be configurable?

[unobtrusive]: https://www.w3.org/wiki/The_principles_of_unobtrusive_JavaScript#:~:text=In%20short%2C%20unobtrusive%20JavaScript%20is,at%20a%20more%20basic%20level. "The principles of unobtrusive JavaScript, by W3C"

### Testing this PR

The [just-the-docs-tests website](https://just-the-docs.github.io/just-the-docs-tests) is currently built using this PR branch.

Browsing the above website, both with and without JavaScript enabled:

1. Check that the navigation links work correctly.
2. Check that the navigation links allow direct navigation to all pages.
3. Check that when previewing the dark color scheme, the nav link to the current page is _not_ highlighted with the background image used for the light scheme.

Note that all collections have to be displayed as _unfolded_ when JavaScript is disabled. **This is a deliberate breaking change** in relation to all releases since v0.4.0.rc1, to make the use of JavaScript in the navigation panel [unobtrusive].

The [profiling results](https://just-the-docs.github.io/just-the-docs-tests/tests/profiles/index/) for building the [just-the-docs-tests website](https://just-the-docs.github.io/just-the-docs-tests) and three larger websites (using this PR as well as some previous releases) indicate that this PR gives significant improvements in build time for larger sites. In principle, the profiling results can be checked using local clones of the sources of the profiled sites and running the provided [makefile](https://github.com/just-the-docs/just-the-docs-tests/blob/main/makefile).
